From 3a71362621e00eba966896b32088dd365ca6c1fa Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Fri, 24 Aug 2018 18:57:29 +0900
Subject: [PATCH 300/909] v4l: vsp1: Disable vsp1 interrupt when startup

commit d88046ebbfb5edaaf726ffea2c7a44814167464c from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

Disable interrupts before request irq. If the interrupt is enabled
and started up without being cleared. the driver may be malfunctioned.
This patch will be added to care for that case.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/vsp1/vsp1_drv.c | 29 +++++++++++++++++++-------
 1 file changed, 22 insertions(+), 7 deletions(-)

diff --git a/drivers/media/platform/vsp1/vsp1_drv.c b/drivers/media/platform/vsp1/vsp1_drv.c
index c13ced37f1ee..c84b81b29f61 100644
--- a/drivers/media/platform/vsp1/vsp1_drv.c
+++ b/drivers/media/platform/vsp1/vsp1_drv.c
@@ -967,13 +967,6 @@ static int vsp1_probe(struct platform_device *pdev)
 		return -EINVAL;
 	}
 
-	ret = devm_request_irq(&pdev->dev, irq->start, vsp1_irq_handler,
-			      IRQF_SHARED, dev_name(&pdev->dev), vsp1);
-	if (ret < 0) {
-		dev_err(&pdev->dev, "failed to request IRQ\n");
-		return ret;
-	}
-
 	/* FCP (optional) */
 	fcp_node = of_parse_phandle(pdev->dev.of_node, "renesas,fcp", 0);
 	if (fcp_node) {
@@ -1028,6 +1021,28 @@ static int vsp1_probe(struct platform_device *pdev)
 
 	dev_dbg(&pdev->dev, "IP version 0x%08x\n", vsp1->version);
 
+	/* Disable interrupts before request irq.
+	 * If the interrupt is not cleared and starts up,
+	 * the driver may be malfunctioned.
+	 */
+	ret = pm_runtime_get_sync(&pdev->dev);
+	if (ret < 0)
+		goto done;
+
+	for (i = 0; i < vsp1->info->lif_count; ++i) {
+		vsp1_write(vsp1, VI6_DISP_IRQ_ENB(i), 0);
+		vsp1_write(vsp1, VI6_WPF_IRQ_ENB(i), 0);
+	}
+
+	pm_runtime_put_sync(&pdev->dev);
+
+	ret = devm_request_irq(&pdev->dev, irq->start, vsp1_irq_handler,
+			       IRQF_SHARED, dev_name(&pdev->dev), vsp1);
+	if (ret < 0) {
+		dev_err(&pdev->dev, "failed to request IRQ\n");
+		goto done;
+	}
+
 	/* Instanciate entities */
 	ret = vsp1_create_entities(vsp1);
 	if (ret < 0) {
-- 
2.17.1

